draft-ietf-quic-version-negotiation.md

```diff
@@ -68,6 +68,75 @@ document are to be interpreted as described in BCP 14 {{!RFC2119}} {{!RFC8174}}
 when, and only when, they appear in all capitals, as shown here.
 
 
+# Server Deployments of QUIC {#server-fleet}
+
+While this document mainly discusses a single QUIC server, it is common for
+deployments of QUIC servers to include a fleet of multiple server instances. We
+therefore define the following terms:
+
+Supported Versions:
+
+: This is the set of versions supported by a given individual server instance.
+
+Fully-Deployed Versions:
+
+: This is the set of QUIC versions that is supported by every single QUIC
+server instance in this deployment.
+
+Partially-Deployed Versions:
+
+: This is the set of versions supported by a given individual server instance,
+but which aren't supported by every single QUIC server instance in this
+deployment.
+
+If a deployment only contains a single server instance, then the Supported
+Versions is equal to the Fully-Deployed Versions, and the Partially-Deployed
+Versions is the empty set. Note however that during software upgrades these
+fields are upgraded at different times, as described below.
+
+Conversely, if a deployment contains multiple server instances, it is possible
+for software updates to not happen at the exact same time on all server
+instances. In this scenario, some instances will have a different Supported
+Versions from other instances in the same fleet. When that happens, a client
+might receive a Version Negotiation packet from a server instance that doesn't
+support some Partially-Supported Versions; in that case, those
+Partially-Supported Versions won't be listed in the Version Negotiation packet.
+
+However, even when there is only a single server instance, it is still possible
+to receive a stale Version Negotiation packet if the server performs its
+software update while the Version Negotiation packet is in flight.
+
+In either of these events, when the client reacts to a Version Negotiation
+packet, it might create a new QUIC connection which could reach a server
+instance which has a different set of Supported Versions than what was received
+in the Version Negotiation packet.
+
+This could cause the version downgrade prevention mechanism described in
+{{downgrade}} to falsely detect a downgrade attack. To avoid that, server
+operators MUST perform a two-step process when they wish to add or remove
+support for a version:
+
+* When adding support for a new version, the first step is to progressively add
+  support for the new version to all server instances. This step updates the
+  Supported Versions but not the Fully-Deployed Versions. Once all server
+  instances have been upgraded, operators wait for at least one minute to allow
+  any in-flight Version Negotiation packets to arrive. Then, the second step is
+  to progressively add the new version to Fully-Deployed Versions on all server
+  instances.
```

Comment:
Do you add the code AND start sending VN with the new version at the same time?  Because sending VN in that state will result in connection failures.  As I said, this probably needs a three step dance.

Reply:
No, the key property here is that VN packets carry the Fully-Deployed Versions, so that happens after the 1min wait. Or am I missing something?

Reply:
So the problem is that if you change VN and Other Versions at the same time. You will have a mismatch between VN and Other Versions in both directions.  Some clients will see VN without the new version and OV with.  Other clients will see VN with the new version and OV without.

It's that former of those two that creates a problem; if the client sees VN without version X and then sees version X in OV, then it will fail the connection.

Reply:
/facepalm. Yup, you're right. Three steps it is.